File: IoTProjectLibrary/CommonNodeCode/routing_table.cpp
#include "routing_table.h"

void initRoutingTable(RoutingTable *table) {
    table->count = 0;
}

void updateRoute(RoutingTable *table, uint8_t neighbor, uint16_t latency, int8_t rssi, uint8_t power_cost) {
    for (int i = 0; i < table->count; i++) {
        if (table->entries[i].neighbor_id == neighbor) {
            table->entries[i].avg_latency = latency;
            table->entries[i].rssi = rssi;
            table->entries[i].power_cost = power_cost;
            return;
        }
    }

    if (table->count < MAX_NEIGHBORS) {
        table->entries[table->count].neighbor_id = neighbor;
        table->entries[table->count].avg_latency = latency;
        table->entries[table->count].rssi = rssi;
        table->entries[table->count].power_cost = power_cost;
        table->count++;
    }
}

uint8_t selectBestNextHop(RoutingTable *table) {
    uint16_t bestCost = 65535;
    uint8_t bestNeighbor = 0xFF;

    for (int i = 0; i < table->count; i++) {
        uint16_t cost = table->entries[i].avg_latency +
                        (table->entries[i].power_cost * 10);

        if (cost < bestCost) {
            bestCost = cost;
            bestNeighbor = table->entries[i].neighbor_id;
        }
    }
    return bestNeighbor;
}

File: py_nodes/CommonNodeCode/routing_table.py
from math import inf

class RouteEntry:
    def __init__(self, neighbor_id, avg_latency, rssi, power_cost):
        self.neighbor_id = neighbor_id
        self.avg_latency = avg_latency
        self.rssi = rssi
        self.power_cost = power_cost

class RoutingTable:
    MAX_NEIGHBOR = 10

    def __init__(self):
        self.entries = {}

    def update_route(self, neighbor_id, avg_latency, rssi, power_cost):
        self.entries[neighbor_id] = RouteEntry(neighbor_id,avg_latency,rssi,power_cost)

        if len(self.entries) > self.MAX_NEIGHBOR:
            oldest_key = next(iter(self.entries))
            del self.entries[oldest_key]

    def select_best_next_hop(self):
        if not self.entries:
            return 0xFF # No neighbors found
        
        best_neighbor = None
        best_cost = float(inf)
        
        for neighbor_id, entry in self.entries.items():
            # Cost = Latency + (PowerCost * 10)
            cost = entry.avg_latency + (entry.power_cost * 10)

            if cost < best_cost:
                best_cost = cost
                best_neighbor = neighbor_id
        return best_neighbor